modify PATH before discovering Python config

kevinushey opened the pull request:
Closes #367.

The fix here effectively moves some of the PATH modification we do to before the attempt to discover the Python configuration + load numpy.

There's a bit of extra orchestration required here as we don't want the PATH changes to persist if we're only attempting to discover Python; ie, we only want to save the PATH changes if we're truly initializing Python. I think I got the semantics right here but it's worth double-checking that I'm doing things correctly.

jjallaire commented on R/virtualenv.R:
```diff
@@ -204,7 +204,7 @@ virtualenv_default_python <- function(python) {
     return(python)
 
-  config <- py_discover_config()
+  config <- py_config()
```
This can't be `py_config()`, because that will actually cause Python to load (and the `virtualenv` family of functions shouldn't do that).

jjallaire commented on R/config.R:
```diff
@@ -299,11 +298,49 @@ python_environments <- function(env_dirs, required_module = NULL) {
+  # we do this up-front in python_config as otherwise attempts to discover
+  # and load numpy can fail, especially on Windows
+  # https://github.com/rstudio/reticulate/issues/367
+  python_home <- python_dirs <- dirname(python)
```
Did we lose the call to `normalizePath(python_dirs)` here?

kevinushey replied:
You're right, I missed that. I'll bring it back.

jjallaire commented on R/package.R:
```diff
@@ -32,16 +33,22 @@ is_python_initialized <- function() {
+    }
+
+    # mark as initializing (so we know to save any PATH changes)
+    .globals$py_initializing <- TRUE
```
`py_disover_config()` can potentially call `py_config()` multiple times, which would in turn cause the `PATH` to be amended multiple times during initialization. I think that rather than using the `py_initializing` flag we should package the PATH manipulating code into a helper function and call it from both places.

kevinushey commented:
I've updated the PR -- let me know if this looks like what you were imagining.

I extracted the path munging code into its own helper function that's a bit smarter about how it munges the PATH, and make sure it's called both during config discovery (and reset after) and also during Python initialization (and not unset if initialization was successful).
